docs/dev/rfcs/query/0007-semi-structured-data-types.md

```diff
@@ -0,0 +1,106 @@
+{"id":7, "date": "1/17", type: "pollop", "score": 3, "poll": 2, "title": "..."}
+{"id":8, "date": "1/18", type: "comment", "parent": 1, "text": "..."}
+```
+
+This column can be split into 10 virtual sub-columns: `tweet.id`, `tweet.date`, `tweet.type`, `tweet.score`, `tweet.desc`, `tweet.title`, `tweet.url`, `tweet.parent`, `tweet.text`, `tweet.poll`.
+The data type of each sub-column can also be automatically deducted from the value, then we can automatically create those sub-columns and insert the corresponding value.
+
+The main advantage of this storage format is that it does not need to parse the raw JSON string when querying the data, which can greatly speed up the query processing.
+The disadvantage is that additional processing is required when inserting data, and the schema of data in each row is not exactly the same. In some scenarios with large differences, many sub-column data will be Null.
+In order to have good performance and balance in various scenarios, we can refer to the optimization algorithms introduced in the paper [JSON Tiles](https://db.in.tum.de/people/sites/durner/papers/json-tiles-sigmod21.pdf).
+
+From the perspective of performance, a better solution is to store data in binary JSON-like format and extract some frequently queried unique keys as sub-columns.
+However, in order to simplify development, we use the JSON format in the first version.
+Binary JSON-like format and separately stored sub-columns will be adopted in a future optimized version.
+
+## Design Details
+
+### Data Types
+
+Add three new values `Variant`, `VariantArray`, `VariantObject` to the enumeration `TypeID`, respectively support these three semi-structured data types.
+Since we now have a value called `Array`, we name the semi-structured `Array` type as `VariantArray` to distinguish from it.
+Define the corresponding structures for these types, and implement the trait `DataType`.
+The `PhysicalTypeID` corresponding to these types are `String`, the JSON value will be converted to a raw string for storage.
+
+```rust
+pub enum TypeID {
+    ...
+    Variant
+    VariantArray
+    VariantObject
+}
+
+pub struct VariantType {}
+
+pub struct VariantArrayType {}
+
+pub struct VariantObjectType {}
+
+```
+
+### Object Column
+
+Currently `Column` is only implemented for fundamental types, custom structs or enumerations like `serde_json::Value` don't have a suitable `Column` implementation to store.
+Define `ObjectColumn` and `MutableObjectColumn` as generic structures to store custom data types, and implement trait `Column` and `MutableColumn` respectively.
+`ObjectType` can be any custom type of structure or enumerations, we can define `JsonColumn` by specified parameter as `serde_json::Value`.
+All the `variant` data will be automatically cast to `serde_json::Value` and generate a `JsonColumn`.
+Other custom data types like BitmapColumn can be supported easily in the future.
+
+```rust
+#[derive(Clone)]
+pub struct ObjectColumn<T: ObjectType> {
```

Comment:
ObjectType can extend from arrow2's `NativeType`.

So values can be `Arc<Bytes<T>>` which will have zero cost to take slice.

Reply:
ok